examples/kaggle_pipeline.py

```diff
@@ -0,0 +1,292 @@
+                validation_fraction=0.1,
+                n_iter_no_change=20,
+                verbose=1,
+            )
+        self.cv_scores = []
+        self.feature_selector = FeatureSelector()
+
+    def _get_tuned_model(self):
+        """Perform hyperparameter tuning using RandomizedSearchCV"""
+        param_distributions = {
+            "learning_rate": uniform(0.01, 0.05),
+            "max_depth": randint(5, 20),
+            "min_samples_leaf": randint(10, 50),
+            "l2_regularization": uniform(0.5, 4.5),
+        }
+
+        base_model = HistGradientBoostingRegressor(
+            max_iter=500, # lowered from 1000
+            early_stopping=True, validation_fraction=0.1, n_iter_no_change=20, random_state=42, verbose=0
+        )
+
+        return RandomizedSearchCV(
+            base_model,
+            param_distributions,
+            n_iter=10,
+            cv=10,
+            scoring="neg_mean_absolute_error",
+            random_state=42,
+            verbose=2,
+        )
+
+    def evaluate_feature_selection_methods(self, X_train, y_train, cv=10):
+        """Compare different feature selection methods"""
+        results = {}
+        methods = {
+            "all_features": lambda x, y: x,
+            "mutual_info": self.feature_selector.mutual_information,
+            "model_based": self.feature_selector.model_based_selection,
+            # "boruta": self.feature_selector.boruta_selection,# intense!
+        }
+
+        for name, method in methods.items():
+            logger.info(f"\nEvaluating {name}...")
+            try:
+                X_selected = method(X_train.copy(), y_train)
+                if self.tune_hyperparameters and hasattr(self.model, "fit"):
+                    logger.info("Performing hyperparameter tuning...")
+                    self.model.fit(X_selected, y_train)
+                    logger.info("Best parameters found:")
+                    logger.info(self.model.best_params_)
+                    logger.info(f"Best cross-validation MAE: {-self.model.best_score_:.4f}")
+
+                scores = self._get_cv_scores(X_selected, y_train, cv)
+                results[name] = {
+                    "n_features": X_selected.shape[1],
+                    "mean_mae": np.mean(scores),
+                    "std_mae": np.std(scores),
+                    "selected_features": list(X_selected.columns),
+                }
+                logger.info(f"Number of features: {X_selected.shape[1]}")
+                logger.info(f"Mean MAE: {np.mean(scores):.4f} (+/- {np.std(scores) * 2:.4f})")
+            except Exception as e:
+                logger.error(f"Error with {name}: {e!s}")
+
+        return results
+
+    def _get_cv_scores(self, X, y, cv):
+        """Helper function to get cross-validation scores"""
+        scores = cross_val_score(
+            self.model,
+            X,
+            y,
+            cv=KFold(n_splits=cv, shuffle=True, random_state=42),
```

Comment:
This could be a source of data leakage, especially if de-tranding has a time-aggregate component. The [TimeSeriesSplit](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.TimeSeriesSplit.html) would probably be better.